Documentation/planning/arcade-powered-source-build/in-arcade.md

```diff
@@ -0,0 +1,132 @@
+restore by adding MSBuild targets. Source-build can add a step to get
+source-build dependencies, with several options:
+
+1. Use transport package. Restores the [intermediate nupkg] and sets it up as a
+   restore source.
+2. Bring-your-own dependencies. A set of args allow the user (or automation) to
+   point to existing intermediates on disk.
+3. Build dependencies from source. Instead of downloading intermediates, clone
+   the repos and build them all.
+
+Any of those options allows the source-build of the current repo to continue
+afterwards with source-built intermediates. Then, source-build sets up
+RestoreSources, a Package Version Props, a blob root, and any repo-specific
+special requirements.
+
+At this point in the build, we should also have all we need to create a
+source-buildable tarball. We should aim to allow the user to create it as early
+as possible and cancel the rest of the build, as they may be a distro maintainer
+who doesn't actually want to build the repo right now. Unfortunately, there are
+some known issues blocking us from creating a source-buildable tarball without
+building the product first, tracked by
+[source-build#831](https://github.com/dotnet/source-build/issues/831). Until we
+find solutions, tarball creation must be done in Publish instead.
+
+### Restore
+
+We need to be careful that no machine-wide caches are used here. Caching a
+source-built nupkg may cause unexpected and incredibly hard to detect errors
+later on.
+
+### Build/Pack
+
+The slew of source-build args normally passed to the build command should be
+moved to Arcade as much as possible. Many of these apply to the Build and Pack
+parts of the build process.
+
+### Publish
+
+This seems like a reasonable place to put [intermediate nupkg] creation. Not all
+repos use this target so care is needed to make sure this doesn't crash.
+
+Creating an intermediate nupkg involves gathering together all shipped artifacts
+(nupkgs and tarballs) into a nupkg. The nupkg the needs to be uploaded to the
+build pipeline along with a manifest to indicate to Arcade how to publish it.
+
+We should also ensure here that no prebuilts were used. This is a validation
+step, however it's essential for source-build health, and quick to check. The
+reason to do this during Publish is that there is no Test phase during a
+source-build and we don't know for sure whether any extra prebuilts were
+restored until the entire build completes. (It's not guaranteed that all
+downloading happens during the Arcade SDK's Restore phase.)
+
+## Supporting a source-built SDK
+
+For bootstrap and N-1 flows, we need to use an SDK from disk. For example, an
+SDK that's restored from an [intermediate nupkg]. The problem is that by the
+time we've restored the intermediate nupkg, we're already inside processes using
+a prebuilt SDK, so we need to launch a new build. A few alternative ways how to
+do this (best first):
```

Comment:
So how I'm thinking of this is we want:
1. Have the option of acquiring a bootstrap SDK, or using a local bootstrap SDK.
2. Using that bootstrap SDK, restore the intermediate nupkg with Arcade, which necessarily brings down a bootstrap Arcade + possibly other packages.
3. Throw everything away except the intermediate nupkg.
4. Start the source-build with the new SDK and packages from the intermediate nupkg.

Is that right?  I'm wondering if we can make the bootstrap flow a normal thing in Arcade, where you could pass a `--with-sdk` argument to `build.sh` if you wanted, or otherwise it would do the normal SDK acquisition, but then for the rest of the toolset restore have an option that just does the dumbest possible package restore and avoid bringing down other things in the first place.

Reply:
Yep, I realized in my prototyping on Core-SDK that we need something like this even just to pull in a source-built Arcade. I'm able to effectively clear out `ProjectsToBuild` in `eng/Build.props`, which seems good enough as lightweight entry point.

I don't think we need to *throw everything away* though. I would rather keep orchestration in MSBuild than bounce between bash scripts and MSBuild, and that may mean we need the bootstrap SDK alive. (It's also nice to avoid constant redownload for devs.)